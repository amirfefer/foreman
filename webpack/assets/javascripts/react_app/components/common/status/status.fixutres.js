export const state = {
  status1: {
    1: {
      id: 1,
      type: 'status1',
      status: 'OK',
    },
    2: {
      id: 1,
      type: 'status2',
      status: 'Error',
      message: 'Error TCP Connection',
    },
  },
  status2: {
    1: {
      id: 1,
      type: 'status2',
      success: true,
      message: { message1: 'message1' },
    },
    2: {
      id: 2,
      type: 'status2',
      success: false,
      message: 'Error Type 1',
    },
    3: {
      id: 3,
      type: 'status2',
      success: false,
      message: {
        warning: { message: 'Warning message' },
      },
    },
  },
};
