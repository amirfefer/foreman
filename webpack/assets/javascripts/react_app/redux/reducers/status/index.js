import Immutable from 'seamless-immutable';
import { STATUS_REQUEST, STATUS_SUCCESS, STATUS_FAILURE } from '../../consts';

const initialState = Immutable({});

export default (state = initialState, action) => {
  const { payload, type } = action;

  switch (type) {
    case STATUS_REQUEST:
      return state;
    case STATUS_SUCCESS:
      return state.setIn([payload.type, payload.id], { ...payload });
    case STATUS_FAILURE:
      return state.setIn([payload.item.type, payload.item.id], { error: payload.error });
    default:
      return state;
  }
};
