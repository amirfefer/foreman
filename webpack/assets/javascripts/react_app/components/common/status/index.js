import React from 'react';
import { connect } from 'react-redux';
import { get } from 'lodash';
import { Icon } from 'patternfly-react';
import { simpleLoader } from '../../common/Loader';
import * as StatusAction from '../../../redux/actions/status/';

export class Status extends React.Component {
  componentDidMount() {
    const { data: { id, type, url }, getStatus } = this.props;
    if (url) {
      getStatus({ id, type }, url);
    }
  }

  render() {
    const {
      status: {
        status, message, error, success,
      },
      getMessage,
    } = this.props;

    if (message && getMessage) {
      return <span>{message[getMessage]}</span>;
    }

    if (status || success !== undefined) {
      if (message && message.warning) {
        return (
          <div title={message.warning.message}>
            <Icon
              type='pf'
              name='warning-triangle-o'
            />
          </div>
        );
      }
      return (
        <div title={typeof message === 'string' ? message : undefined}>
          <Icon
            type='pf'
            name={status === 'OK' || success ? 'ok' : 'error-circle-o'}
          />
        </div>
      );
    }

    if (error) {
      return (
        <div title={message}>
          <Icon
            type='pf'
            name='error-circle-o' />
        </div>
      );
    }

    return simpleLoader('xs');
  }
}

const mapStateToProps = (state, ownProps) => {
  const { type, id } = ownProps.data;
  return {
    status: get(state.status, `${type}.${id}`) || {},
  };
};

export default connect(mapStateToProps, StatusAction)(Status);
