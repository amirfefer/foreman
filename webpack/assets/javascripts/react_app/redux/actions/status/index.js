import { STATUS_REQUEST, STATUS_SUCCESS, STATUS_FAILURE } from '../../consts';
import { ajaxRequestAction } from '../common';

export const getStatus = ({ id, type }, url) => dispatch =>
  ajaxRequestAction({
    dispatch,
    requestAction: STATUS_REQUEST,
    successAction: STATUS_SUCCESS,
    failedAction: STATUS_FAILURE,
    url,
    item: { id, type },
  });
