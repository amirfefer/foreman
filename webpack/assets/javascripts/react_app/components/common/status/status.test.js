import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Status } from './';
import { state } from './status.fixutres';

describe('Status', () => {
  it('status 1 should render ok icon', () => {
    const wrapper = shallow(<Status data={{ type: 'status1', id: 1 }} status={state.status1[1]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('status 1 should render error icon', () => {
    const wrapper = shallow(<Status data={{ type: 'status1', id: 2 }} status={state.status1[2]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('status 2 should render ok icon', () => {
    const wrapper = shallow(<Status data={{ type: 'status2', id: 1 }} status={state.status2[1]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('status 2 should render error icon', () => {
    const wrapper = shallow(<Status data={{ type: 'status2', id: 2 }} status={state.status2[2]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('status 2 should render warning icon', () => {
    const wrapper = shallow(<Status data={{ type: 'status2', id: 3 }} status={state.status2[3]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render a message', () => {
    const wrapper = shallow(<Status
        getMessage="message1"
        data={{ type: 'status2', id: 2 }}
        status={state.status2[1]}
      />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
