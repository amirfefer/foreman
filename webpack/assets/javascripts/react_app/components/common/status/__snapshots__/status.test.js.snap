// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status should render a message 1`] = `
<span>
  message1
</span>
`;

exports[`Status status 1 should render error icon 1`] = `
<div
  title="Error TCP Connection"
>
  <Icon
    name="error-circle-o"
    type="pf"
  />
</div>
`;

exports[`Status status 1 should render ok icon 1`] = `
<div>
  <Icon
    name="ok"
    type="pf"
  />
</div>
`;

exports[`Status status 2 should render error icon 1`] = `
<div
  title="Error Type 1"
>
  <Icon
    name="error-circle-o"
    type="pf"
  />
</div>
`;

exports[`Status status 2 should render ok icon 1`] = `
<div>
  <Icon
    name="ok"
    type="pf"
  />
</div>
`;

exports[`Status status 2 should render warning icon 1`] = `
<div
  title="Warning message"
>
  <Icon
    name="warning-triangle-o"
    type="pf"
  />
</div>
`;
